File: lib/python-sdk/examples/grants_gov_transforms.py
#!/usr/bin/env python3
"""Round-trip validation using the cg-grants-gov plugin.

Fetches a real opportunity from Simpler.Grants.gov, transforms it to
CommonGrants format via ``to_common``, then back to the source format via
``from_common``, and validates that key fields were preserved.


Run with: poetry run python examples/grants_gov_transforms.py <opportunityId>
          or set GRANTS_GOV_OPP_ID and omit the argument.
"""

from __future__ import annotations

import json
import os
import sys
from typing import Any

from cg_grants_gov import GrantsGovOpportunitySchema, grants_gov


from common_grants_sdk.client import Client
from common_grants_sdk.client.config import Config

# =============================================================================
# Config
# =============================================================================

opp_id = sys.argv[1] if len(sys.argv) > 1 else os.environ.get("GRANTS_GOV_OPP_ID")
api_key = "two_org_user_key"
base_url = "http://localhost:8080"
config = Config(
    base_url=base_url,
    api_key="two_org_user_key",
    timeout=5.0,
)
client = Client(config)

# =============================================================================
# Helpers
# =============================================================================


def check(label: str, condition: bool, detail: str = "") -> None:
    if condition:
        print(f"  PASS  {label}")
    else:
        suffix = f"\n        {detail}" if detail else ""
        print(f"  FAIL  {label}{suffix}", file=sys.stderr)
        sys.exit(1)


def to_json(obj: Any) -> str:
    """Serialize a Pydantic model or plain dict to pretty-printed JSON."""
    if hasattr(obj, "model_dump"):
        data = obj.model_dump(mode="json")
    else:
        data = obj
    return json.dumps(data, indent=2, default=str)


def deep_diff(
    a: Any,
    b: Any,
    path: str = "",
    diffs: list[tuple[str, Any, Any]] | None = None,
) -> list[tuple[str, Any, Any]]:
    """Recursively collect (path, a_val, b_val) for every leaf that differs."""
    if diffs is None:
        diffs = []
    if isinstance(a, dict) and isinstance(b, dict):
        all_keys = set(a) | set(b)
        for k in sorted(all_keys):
            deep_diff(a.get(k), b.get(k), f"{path}.{k}" if path else k, diffs)
    elif isinstance(a, list) and isinstance(b, list):
        for i, (av, bv) in enumerate(zip(a, b)):
            deep_diff(av, bv, f"{path}[{i}]", diffs)
        if len(a) != len(b):
            diffs.append((f"{path}[len]", len(a), len(b)))
    else:
        if a != b:
            diffs.append((path, a, b))
    return diffs


# =============================================================================
# Main
# =============================================================================


def main() -> None:
    print("=== Grants.gov Round-Trip Transform ===")
    print(f"Opportunity: {opp_id}")
    print(f"Base URL:    {base_url}\n")

    # 1. Fetch raw opportunity from Simpler.Grants.gov
    print("Fetching opportunity...")
    url = f"/v1/opportunities/{opp_id}"
    response = client.get(url)
    response.raise_for_status()
    body = response.json()
    print("  fetched OK\n")
    print(f"=== [1] RAW API RESPONSE ===\n{to_json(body['data'])}\n")

    # 2. Validate raw response shape
    print("Validating raw response...")
    raw = GrantsGovOpportunitySchema.model_validate(body["data"])
    print("  validated OK\n")

    # 3. to_common
    print("Running to_common...")
    tr1 = grants_gov.schemas.Opportunity.to_common(raw)  # type: ignore[attr-defined]
    if tr1.errors:
        errs = "; ".join(f"[{e.path or '?'}] {e}" for e in tr1.errors)
        print(f"  FAIL  to_common produced errors: {errs}", file=sys.stderr)
        sys.exit(1)
    common = tr1.result
    print("  to_common OK\n")
    print(f"=== [2] COMMON (to_common output) ===\n{to_json(common)}\n")
    # 4. from_common
    print("Running from_common...")
    tr2 = grants_gov.schemas.Opportunity.from_common(common)  # type: ignore[attr-defined]
    if tr2.errors:
        errs = "; ".join(f"[{e.path or '?'}] {e}" for e in tr2.errors)
        print(f"  FAIL  from_common produced errors: {errs}", file=sys.stderr)
        sys.exit(1)
    back = tr2.result
    print("  from_common OK\n")
    print(f"=== [3] BACK (from_common output) ===\n{to_json(back)}\n")

    # 5. Validate round-trip field preservation
    print("Validating round-trip field preservation...")

    # --- Exact match ---
    check(
        "opportunity_id round-trips",
        str(back.opportunity_id) == str(raw.opportunity_id),
        f"expected {raw.opportunity_id}, got {back.opportunity_id}",
    )

    check(
        "opportunity_number round-trips",
        back.opportunity_number == raw.opportunity_number,
        f"expected {raw.opportunity_number!r}, got {back.opportunity_number!r}",
    )

    check(
        "agency_code round-trips",
        back.agency_code == raw.agency_code,
        f"expected {raw.agency_code!r}, got {back.agency_code!r}",
    )

    check(
        "agency_name round-trips",
        back.agency_name == raw.agency_name,
        f"expected {raw.agency_name!r}, got {back.agency_name!r}",
    )

    # --- Semantic match ---
    # Status: posted→open→posted, archived→closed→closed, forecasted→forecasted→forecasted
    expected_status: dict[str, str] = {
        "posted": "posted",
        "archived": "closed",
        "forecasted": "forecasted",
        "closed": "closed",
    }
    expected_back_status = expected_status.get(raw.opportunity_status, "posted")
    check(
        f"opportunity_status semantic ({raw.opportunity_status} → ... → {expected_back_status})",
        back.opportunity_status == expected_back_status,
        f"expected {expected_back_status!r}, got {back.opportunity_status!r}",
    )

    # Funding amounts (skip if null in source)
    raw_floor = raw.summary.award_floor if raw.summary else None
    raw_ceiling = raw.summary.award_ceiling if raw.summary else None
    back_floor = back.summary.award_floor if back.summary else None
    back_ceiling = back.summary.award_ceiling if back.summary else None

    if raw_floor is not None:
        check(
            f"award_floor round-trips ({raw_floor})",
            back_floor == raw_floor,
            f"expected {raw_floor}, got {back_floor}",
        )
    else:
        print("  PASS  award_floor (null in source — skipped)")

    if raw_ceiling is not None:
        check(
            f"award_ceiling round-trips ({raw_ceiling})",
            back_ceiling == raw_ceiling,
            f"expected {raw_ceiling}, got {back_ceiling}",
        )
    else:
        print("  PASS  award_ceiling (null in source — skipped)")

    # --- Presence check ---
    check(
        "opportunity_title preserved",
        back.opportunity_title == raw.opportunity_title,
        f"expected {raw.opportunity_title!r}, got {back.opportunity_title!r}",
    )

    check(
        "created_at non-null",
        back.created_at is not None,
        f"got: {back.created_at!r}",
    )

    check(
        "updated_at non-null",
        back.updated_at is not None,
        f"got: {back.updated_at!r}",
    )

    raw_listing_count = len(raw.opportunity_assistance_listings)
    back_listing_count = len(back.opportunity_assistance_listings)
    check(
        "assistance_listings count preserved",
        back_listing_count == raw_listing_count,
        f"expected {raw_listing_count}, got {back_listing_count}",
    )

    # --- Full deep comparison: raw vs back ---
    print("\nRunning deep diff (raw vs from_common)...")
    raw_dict = raw.model_dump(mode="json")
    back_dict = back.model_dump(mode="json")
    diffs = deep_diff(raw_dict, back_dict)
    if diffs:
        print(f"  {len(diffs)} field(s) differ after round-trip:")
        for path, a_val, b_val in diffs:
            print(f"    {path}")
            print(f"      raw:  {a_val!r}")
            print(f"      back: {b_val!r}")
    else:
        print("  PASS  all fields identical after round-trip")

    print("\n=== All checks passed ===")


if __name__ == "__main__":
    main()

File: lib/ts-sdk/examples/grants-gov-transforms.ts
/**
 * Round-trip validation using the @common-grants/cg-grants-gov plugin.
 *
 * Fetches a real opportunity from Simpler.Grants.gov, transforms it to
 * CommonGrants format via `toCommon`, then back to the source format via
 * `fromCommon`, and validates that key fields were preserved.
 *
 * Run with: pnpm example:grants-gov [opportunityId]
 *   or set GRANTS_GOV_OPP_ID and run: pnpm example:grants-gov
 */

import plugin, { GrantsGovOpportunitySchema } from "@common-grants/cg-grants-gov";
import { Client, Auth } from "@common-grants/sdk/client";

// ============================================================================
// Config
// ============================================================================

const oppId = process.argv[2] ?? process.env.GRANTS_GOV_OPP_ID;
const apiKey = "two_org_user_key";
const baseUrl = "http://localhost:8080";
const config = {
  baseUrl,
  auth: Auth.apiKey(apiKey),
  timeout: 5000,
};
const client = new Client(config);

if (!oppId) {
  console.error("Usage: pnpm example:grants-gov <opportunityId>");
  console.error("       or set GRANTS_GOV_OPP_ID env var");
  process.exit(1);
}

// ============================================================================
// Helpers
// ============================================================================

function fail(label: string, detail?: string): never {
  console.error(`  FAIL  ${label}${detail ? `\n        ${detail}` : ""}`);
  process.exit(1);
}

function pass(label: string): void {
  console.log(`  PASS  ${label}`);
}

function check(label: string, condition: boolean, detail?: string): void {
  if (!condition) fail(label, detail);
  pass(label);
}

function toJson(obj: unknown): string {
  return JSON.stringify(obj, null, 2);
}

type Diff = { path: string; raw: unknown; back: unknown };

function deepDiff(a: unknown, b: unknown, path = "", diffs: Diff[] = []): Diff[] {
  if (
    a !== null &&
    b !== null &&
    typeof a === "object" &&
    typeof b === "object" &&
    !Array.isArray(a) &&
    !Array.isArray(b)
  ) {
    const aObj = a as Record<string, unknown>;
    const bObj = b as Record<string, unknown>;
    const keys = new Set([...Object.keys(aObj), ...Object.keys(bObj)]);
    for (const k of [...keys].sort()) {
      deepDiff(aObj[k], bObj[k], path ? `${path}.${k}` : k, diffs);
    }
  } else if (Array.isArray(a) && Array.isArray(b)) {
    const len = Math.max(a.length, b.length);
    for (let i = 0; i < len; i++) {
      deepDiff(a[i], b[i], `${path}[${i}]`, diffs);
    }
    if (a.length !== b.length) {
      diffs.push({ path: `${path}[len]`, raw: a.length, back: b.length });
    }
  } else if (a !== b) {
    diffs.push({ path, raw: a, back: b });
  }
  return diffs;
}

// ============================================================================
// Main
// ============================================================================

async function main(): Promise<void> {
  console.log(`=== Grants.gov Round-Trip Transform ===`);
  console.log(`Opportunity: ${oppId}`);
  console.log(`Base URL:    ${baseUrl}\n`);

  // 1. Fetch raw opportunity from Simpler.Grants.gov
  console.log("Fetching opportunity...");
  const path = `/v1/opportunities/${oppId}`;
  const response = await client.get(path);

  if (!response.ok) {
    fail(`HTTP ${response.status} from ${path}`, await response.text());
  }

  const body = (await response.json()) as { data: unknown };
  console.log("  fetched OK\n");
  console.log(`=== [1] RAW API RESPONSE ===\n${toJson(body.data)}\n`);

  // 2. Validate raw response shape
  console.log("Validating raw response...");
  const parseResult = GrantsGovOpportunitySchema.safeParse(body.data);
  if (!parseResult.success) {
    fail("GrantsGovOpportunitySchema.parse(body.data)", parseResult.error.message);
  }
  const raw = parseResult.data;
  console.log("  validated OK\n");

  // 3. toCommon
  console.log("Running toCommon...");
  const { result: common, errors: toErrors } = plugin.schemas.Opportunity.toCommon(raw);
  if (toErrors.length > 0) {
    fail(
      "toCommon produced errors",
      toErrors.map(e => `[${e.path ?? "?"}] ${e.message}`).join("; ")
    );
  }
  console.log("  toCommon OK\n");
  console.log(`=== [2] COMMON (toCommon output) ===\n${toJson(common)}\n`);

  // 4. fromCommon
  console.log("Running fromCommon...");
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const { result: back, errors: fromErrors } = plugin.schemas.Opportunity.fromCommon(common as any);
  if (fromErrors.length > 0) {
    fail(
      "fromCommon produced errors",
      fromErrors.map(e => `[${e.path ?? "?"}] ${e.message}`).join("; ")
    );
  }
  console.log("  fromCommon OK\n");
  console.log(`=== [3] BACK (fromCommon output) ===\n${toJson(back)}\n`);

  // 5. Validate round-trip field preservation
  console.log("Validating round-trip field preservation...");

  // --- Exact match ---
  check(
    "opportunity_id round-trips",
    back.opportunity_id === raw.opportunity_id,
    `expected ${raw.opportunity_id}, got ${back.opportunity_id}`
  );

  check(
    "opportunity_number round-trips",
    back.opportunity_number === raw.opportunity_number,
    `expected ${raw.opportunity_number}, got ${back.opportunity_number}`
  );

  check(
    "agency_code round-trips",
    back.agency_code === raw.agency_code,
    `expected ${raw.agency_code}, got ${back.agency_code}`
  );

  check(
    "agency_name round-trips",
    back.agency_name === raw.agency_name,
    `expected ${raw.agency_name}, got ${back.agency_name}`
  );

  // --- Semantic match ---
  // Status: posted→open→posted, archived→closed→closed, forecasted→forecasted→forecasted
  const expectedStatus: Record<string, string> = {
    posted: "posted",
    archived: "closed",
    forecasted: "forecasted",
    closed: "closed",
  };
  const expectedBack = expectedStatus[raw.opportunity_status] ?? "posted";
  check(
    `opportunity_status semantic (${raw.opportunity_status} → ... → ${expectedBack})`,
    back.opportunity_status === expectedBack,
    `expected ${expectedBack}, got ${back.opportunity_status}`
  );

  // Funding amounts (skip if null in source)
  const rawFloor = raw.summary?.award_floor;
  const rawCeiling = raw.summary?.award_ceiling;

  if (rawFloor != null) {
    check(
      `award_floor round-trips (${rawFloor})`,
      back.summary?.award_floor === rawFloor,
      `expected ${rawFloor}, got ${back.summary?.award_floor}`
    );
  } else {
    pass("award_floor (null in source — skipped)");
  }

  if (rawCeiling != null) {
    check(
      `award_ceiling round-trips (${rawCeiling})`,
      back.summary?.award_ceiling === rawCeiling,
      `expected ${rawCeiling}, got ${back.summary?.award_ceiling}`
    );
  } else {
    pass("award_ceiling (null in source — skipped)");
  }

  // --- Presence check ---
  check(
    "opportunity_title preserved",
    back.opportunity_title === raw.opportunity_title,
    `expected ${raw.opportunity_title}, got ${back.opportunity_title}`
  );

  check(
    "created_at non-null",
    typeof back.created_at === "string" && back.created_at.length > 0,
    `got: ${back.created_at}`
  );

  check(
    "updated_at non-null",
    typeof back.updated_at === "string" && back.updated_at.length > 0,
    `got: ${back.updated_at}`
  );

  check(
    "assistance_listings count preserved",
    (back.opportunity_assistance_listings?.length ?? 0) ===
      (raw.opportunity_assistance_listings?.length ?? 0),
    `expected ${raw.opportunity_assistance_listings?.length ?? 0}, got ${back.opportunity_assistance_listings?.length ?? 0}`
  );

  // --- Full deep comparison: raw vs back ---
  console.log("\nRunning deepDiff (raw vs fromCommon)...");
  const diffs = deepDiff(raw, back);
  if (diffs.length > 0) {
    console.log(`  ${diffs.length} field(s) differ after round-trip:`);
    for (const { path, raw: aVal, back: bVal } of diffs) {
      console.log(`    ${path}`);
      console.log(`      raw:  ${JSON.stringify(aVal)}`);
      console.log(`      back: ${JSON.stringify(bVal)}`);
    }
  } else {
    pass("all fields identical after round-trip");
  }

  console.log("\n=== All checks passed ===");
}

main().catch(err => {
  console.error("Unexpected error:", err);
  process.exit(1);
});
